docs/developer-guide.md

```diff
@@ -0,0 +1,297 @@
+
+Uncomment `grafana` and `prometheus` lines in `risedev.yml` to enable Grafana and Prometheus services. 
+
+### Tracing
+
+Compute nodes support streaming tracing. Tracing is not enabled by default. You will need to
+use `./risedev configure` to enable tracing components. After that, simply uncomment `jaeger`
+service in `risedev.yml`.
+
+### Dashboard
+
+You may use RisingWave Dashboard to see actors in the system. It will be started along with meta node.
+
+### Logging
+
+The Rust components use `tokio-tracing` to handle both logging and tracing. The default log level is set as:
+
+* Third-party libraries: warn
+* Other libraries: debug
+
+If you need to adjust log levels, change the logging filters in `utils/logging/lib.rs`.
+
+
+## Test your code changes
+
+Before you submit a PR, we recommend that you fully test the code changes and perform necessary checks.
+
+The RisingWave project enforces several checks in CI. Every time the code is modified, you need to perform the checks and ensure they pass.
+
+### Lint
+
+RisingWave requires all code to pass fmt, clippy, sort and hakari checks. Run the following commands to install test tools and perform these checks.
+
+```shell
+./risedev install-tools # Install required tools for running unit tests
+./risedev c             # Run all checks. Shortcut for ./risedev check
+```
+
+### Unit tests
+
+RiseDev runs unit tests with cargo-nextest. To run unit tests:
+
+```shell
+./risedev install-tools # Install required tools for running unit tests
+./risedev test          # Run unit tests
+```
+
+If you want to see the coverage report, run this command:
+
+```shell
+./risedev test-cov
+```
+
+### Planner tests
+
+RisingWave's SQL frontend has SQL planner tests. For more information, see [Planner Test Guide](src/frontend/test_runner/README.md).
+
+### End-to-end tests
+
+Currently, we use [sqllogictest-rs](https://github.com/risinglightdb/sqllogictest-rs) to run RisingWave e2e tests.
+
+sqllogictest installation is included when you install test tools with the `./risedev install-tools` command. You may also install it with:
+
+```shell
+cargo install --git https://github.com/risinglightdb/sqllogictest-rs --features bin
+```
+
+Before running end-to-end tests, you will need to start a full cluster first:
+
+```shell
+./risedev d
+```
+
+Then run the end-to-end tests (replace `**/*.slt` with the test case directories and files available):
+
+```shell
+./risedev slt -p 4566 './e2e_test/streaming/**/*.slt'
+```
+
+After running e2e tests, you may kill the cluster and clean data.
+
+```shell
+./risedev k  # shortcut for ./risedev kill
+./risedev clean-data
+```
+
+RisingWave's codebase is constantly changing. The persistent data might not be stable. In case of unexpected decode errors, try `./risedev clean-data` first.
+
+### End-to-end tests on CI
+
+Basically, CI is using the following two configurations to run the full e2e test suite:
+
+```shell
+./risedev dev ci-3cn-1fe
+./risedev dev ci-1cn-1fe
```

Comment:
This config is not used anymore. Just retain `./risedev dev ci-3cn-1fe` here.

Reply:
cool.